File: cc/056.cc
// Project Euler Problem 056
//
// Considering natural numbers of the form, a^b, where a, b < 100,
// what is the maximum digital sum?

#include <cstdio>
#include <vector>

#include "bigmath.h"

int main(int argc, const char* argv[]) {
  int max = 0;
  for (int a=1; a < 100; a++) {
    vector<int> vec;
    vec.push_back(1);
    times_digits_reversed(vec, a);
    for (int b=1; b < 100; b++) {
      vector<int> vec2 = vec;
      pow_digits_reversed(vec2, b);
      int sum = sum_digits(vec2);
      if (sum > max) max = sum;
    }
  }
  printf("%d\n", max);
}

File: go/problems/056.go
// Project Euler Problem 056
//
// Considering natural numbers of the form, a^b, where a, b < 100,
// what is the maximum digital sum?

package problems

import (
	"fmt"
	"math/big"
)

func bigDigitSum(n *big.Int) int {
	sum := 0
	for _, c := range n.String() {
		sum += int(c - '0')
	}
	return sum
}

func Problem056() string {
	max := 0
	for a := int64(2); a < 100; a++ {
		bigA := big.NewInt(a)
		for b := int64(2); b < 100; b++ {
			bigB := big.NewInt(b)
			z := big.NewInt(0)
			z.Exp(bigA, bigB, nil)
			sum := bigDigitSum(z)
			if sum > max {
				max = sum
			}
		}
	}
	return fmt.Sprint(max)
}

func init() {
	Functions[56] = Problem056
}
